plugins/outputs/firehose/firehose.go

```diff
@@ -0,0 +1,268 @@
+
+	// we simply create the skeleton here. AWS doesn't attempt
+	// any connection yet, so we don't know if an error will happen.
+	svc := firehose.New(configProvider)
+	f.svc = svc
+	return nil
+}
+
+func (f *FirehoseOutput) Close() error {
+	return nil
+}
+
+func (f *FirehoseOutput) SetSerializer(serializer serializers.Serializer) {
+	f.serializer = serializer
+}
+
+// writeToFirehose uses the AWS GO SDK to write batched records to firehose and
+// queue the failed writes to the errorBuffer of f
+func writeToFirehose(f *FirehoseOutput, b [][]byte, submitAttemptCount int64) {
+	start := time.Now()
+	dataSize := 0 // track the size of the record in bytes
+	var values []byte
+
+	// take metrics and flatten into a record
+
+	if f.EnableGzipCompression == true {
+		var buf bytes.Buffer
+		zw := gzip.NewWriter(&buf)
+
+		for _, line := range b {
+			_, err := zw.Write(line)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+
+		if err := zw.Close(); err != nil {
+			log.Fatal(err)
+		}
+
+		values = buf.Bytes()
+		dataSize = len(values)
+	} else {
+		for _, line := range b {
+			values = append(values, line...)
+			dataSize = dataSize + len(line)
+		}
+	}
+
+	// todo.. check that we are not exceeding the max record size of 1MB to firehose
+	// though.. we'd be really impressed if any single telegraf metric write is >1MB. That's a lot.
+	// for now, i've updated the docstring to ask that users adjust metric_batch_size and flush_interval
+	// if they run into this.  A proper solution here would need to predict and (if needed) move lines
+	// around to get as close to 1MB as possible.. the level of effort is high enough to postpone for now.
+
+	d := firehose.Record{
+		Data: values,
+	}
+
+	r := []*firehose.Record{}
+	r = append(r, &d)
+
+	batchInput := &firehose.PutRecordBatchInput{}
+	batchInput.SetDeliveryStreamName(f.DeliveryStreamName)
+	batchInput.SetRecords(r)
+
+	// attempt to send data to firehose
+	resp, err := f.svc.PutRecordBatch(batchInput)
+
+	// if we had a failure, log it and enqueue the request for next time
+	if (aws.Int64Value(resp.FailedPutCount) > 0) || (err != nil) {
+		if err != nil {
+			log.Printf("E! firehose: Unable to write to Firehose : %+v \n", err.Error())
+		}
+		// Log the RequestResponse if there were any failed Puts (in the batch Put)
+		if aws.Int64Value(resp.FailedPutCount) > 0 {
+			log.Printf("E! firehose: One or more Put operations failed : %+v \n", resp)
+		}
+		newErrorEntry := errorEntry{submitAttemptCount: submitAttemptCount + 1, batch: b}
+		f.errorBuffer = append(f.errorBuffer, &newErrorEntry)
+		f.totalErrorCount += 1
+		return
+	}
+	log.Printf("I! firehose: successfully sent %d metric batches in %+v. Total size: %d bytes.\n", len(r), time.Since(start), dataSize)
+}
+
+// Write function is responsible for first writing the batch entries held in
+// the errorBuffer of f and then writing metrics in batches of 500.
+func (f *FirehoseOutput) Write(metrics []telegraf.Metric) error {
+
+	if len(metrics) == 0 {
```

Comment:
You can remove this check, as it is done right before this function is called.  That said, below you may want to check if the `prepared_batch` is empty.